In [1]:
import re
import pandas as pd
import string
import nltk
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
#import nltk
nltk.download('punkt')
nltk.download('stopwords')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\RPC\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\RPC\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [2]:
from nltk.tokenize import word_tokenize

In [3]:
dataset = pd.read_csv('youtoxic_english.csv')
dataset

CommentId      VideoId  \
0    Ugg2KwwX0V8-aXgCoAEC  04kJtp6pVXI   
1    Ugg2s5AzSPioEXgCoAEC  04kJtp6pVXI   
2    Ugg3dWTOxryFfHgCoAEC  04kJtp6pVXI   
3    Ugg7Gd006w1MPngCoAEC  04kJtp6pVXI   
4    Ugg8FfTbbNF8IngCoAEC  04kJtp6pVXI   
..                    ...          ...   
995  Ugi5ADt10EdDz3gCoAEC  XRuCW80L9mA   
996  Ugifh2DMhBbDkHgCoAEC  XRuCW80L9mA   
997  Ugj_plbGBjjzYXgCoAEC  XRuCW80L9mA   
998  Ugj0bah1De8xy3gCoAEC  XRuCW80L9mA   
999  UgjBJKQSoQMQ6ngCoAEC  XRuCW80L9mA   

                                                  Text  IsToxic  IsAbusive  \
0    If only people would just take a step back and...    False      False   
1    Law enforcement is not trained to shoot to app...     True       True   
2    \nDont you reckon them 'black lives matter' ba...     True       True   
3    There are a very large number of people who do...    False      False   
4    The Arab dude is absolutely right, he should h...    False      False   
..                                                 ...      ...        ...   
995  I remember that they sent in the national defe...    False      False   
996  Stats don`t represent the problem. Race baitin...     True      False   
997  The quote from the mother... Wow that hit hard...    False      False   
998                            this video is so racist    False      False   
999      God, the narrator has such an annoying lisp.     False      False   

     IsThreat  IsProvocative  IsObscene  IsHatespeech  IsRacist  \
0       False          False      False         False     False   
1       False          False      False         False     False   
2       False          False       True         False     False   
3       False          False      False         False     False   
4       False          False      False         False     False   
..        ...            ...        ...           ...       ...   
995     False          False      False         False     False   
996     False          False      False          True      True   
997     False          False      False         False     False   
998     False          False      False         False     False   
999     False          False      False         False     False   

     IsNationalist  IsSexist  IsHomophobic  IsReligiousHate  IsRadicalism  
0            False     False         False            False         False  
1            False     False         False            False         False  
2            False     False         False            False         False  
3            False     False         False            False         False  
4            False     False         False            False         False  
..             ...       ...           ...              ...           ...  
995          False     False         False            False         False  
996          False     False         False            False         False  
997          False     False         False            False         False  
998          False     False         False            False         False  
999          False     False         False            False         False  

[1000 rows x 15 columns]

In [4]:

# Sélection de la deuxième colonne
col2 = dataset.iloc[:, 2]

# Affichage du tableau
print(col2)


0      If only people would just take a step back and...
1      Law enforcement is not trained to shoot to app...
2      \nDont you reckon them 'black lives matter' ba...
3      There are a very large number of people who do...
4      The Arab dude is absolutely right, he should h...
                             ...                        
995    I remember that they sent in the national defe...
996    Stats don`t represent the problem. Race baitin...
997    The quote from the mother... Wow that hit hard...
998                              this video is so racist
999        God, the narrator has such an annoying lisp. 
Name: Text, Length: 1000, dtype: object


In [5]:


# Fonction de nettoyage de texte
def clean_text(text):
    # Convertir le texte en minuscules
    text = text.lower()
    
    # Retirer la ponctuation
    text = re.sub('[%s]' % re.escape(string.punctuation), '', text)
    
    # Retirer les chiffres
    text = re.sub('\d+', '', text)
    
    # Retirer les stopwords
    stopwords_list = stopwords.words('english')
    tokens = nltk.word_tokenize(text)
    tokens = [word for word in tokens if word not in stopwords_list]
    text = ' '.join(tokens)
    
    # Stemming des mots
    stemmer = PorterStemmer()
    tokens = nltk.word_tokenize(text)
    tokens = [stemmer.stem(word) for word in tokens]
    text = ' '.join(tokens)
    
    return text

# Appliquer la fonction de nettoyage à la deuxième colonne
col2_clean = col2.apply(lambda x: clean_text(x))

# Afficher le résultat
print(col2_clean)


0      peopl would take step back make case wasnt any...
1      law enforc train shoot apprehend train shoot k...
2      dont reckon black live matter banner held whit...
3      larg number peopl like polic offic call crimin...
4      arab dude absolut right shot extra time shoot ...
                             ...                        
995                            rememb sent nation defenc
996    stat dont repres problem race bait attitud who...
997                       quot mother wow hit hard accur
998                                         video racist
999                                god narrat annoy lisp
Name: Text, Length: 1000, dtype: object


In [6]:
pip install demoji

     -------------------------------------- 42.9/42.9 kB 693.9 kB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.


In [7]:
import demoji

def remove_emojis(text):
    # Remove emojis from text
    text_without_emojis = demoji.replace(text, "")
    return text_without_emojis

In [9]:
from collections import defaultdict
import math

# Exemple d'ensemble de commentaires


# Initialisation du dictionnaire de fréquences
frequences = defaultdict(int)

# Prétraitement des commentaires et calcul des fréquences
for commentaire in col2_clean:
    # Prétraitement : suppression de la ponctuation, mise en minuscules, etc.

    # Diviser le commentaire en mots individuels
    mots = commentaire.split()

    # Calcul des fréquences des mots
    for mot in mots:
        frequences[mot] += 1

# Affichage des fréquences des mots
# for mot, frequence in frequences.items():
#     print(mot, ":", frequence)
# Calcul de l'IDF pour chaque terme
N = len(col2_clean)
idf = {}
for mot, frequence in frequences.items():
    idf[mot] = math.log(N / (frequence + 1))

# Calcul de la pondération TF-IDF pour chaque terme dans chaque commentaire
for commentaire in col2_clean:
    tfidf = defaultdict(float)
    mots = commentaire.split()
    for mot in mots:
        tf = frequences[mot] / len(mots)
        tfidf[mot] = tf * idf[mot]

    # Affichage des pondérations TF-IDF pour le commentaire
    print(commentaire)
    for mot, poids in tfidf.items():
        print(mot, ":", poids)
    print()

peopl would take step back make case wasnt anyon except two peopl situat lump mess take matter hand make kind protest selfish without ration thought investig guy video heavili emot hype want heard get heard press never reason discuss kudo smerconish keep level whole time let masri make fool dare tore citi protest make dishonor entir incid hate way sinc polic brutal becom epidem wish everyon would stop pretend like knew exactli go there measur amount peopl honestli wit incid none us clue way whole issu swung grand juri inform trust major rule right cours action let also thank polic offic america actual serv protect even your bit jerk pull respect job know someon mani peopl go pout held account action peopl hate polic need offic two around emerg
peopl : 2.7072950901663204
would : 1.9646183566375777
take : 1.1222083649754457
step : 0.29212593498579337
back : 1.1379138093344616
make : 1.522717649818031
case : 0.937955483817541
wasnt : 0.26200152062880705
anyon : 0.643331344086556
except : 

rogan : 3.755355129012901
tell : 12.128880773389001
good : 15.39556941246521
cop : 31.997154124283668
put : 13.18734946535165
bad : 9.53145925762071
one : 23.81190657444629
jail : 5.405703880126167

agre joe cop extrem difficult job mani qualifi handl ye lot good cop howev intellig unplug matrix joe alway claim seem complet unawar ignor realiti countri corrupt prison industri complex tie fail judici system well racial agenda behind polic brutal
agre : 1.4698967772579745
joe : 1.0361632918473207
cop : 7.199359677963826
extrem : 0.9421061403291836
difficult : 0.1553652024605548
job : 2.849546134868815
mani : 2.6675650165413685
qualifi : 0.1553652024605548
handl : 1.3028417764561802
ye : 1.1274650015459415
lot : 1.7834923349170635
good : 3.4640031178046726
howev : 0.9421061403291836
intellig : 0.8449549040279027
unplug : 0.1553652024605548
matrix : 0.2904571495157014
alway : 2.14457833296466
claim : 1.2162833730283877
seem : 2.2133276724683615
complet : 1.5506874587783837
unawar : 0.15536

# clustring avec l'algorithme k-means

In [57]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.cluster import KMeans
import pandas as pd

# Calcul des vecteurs TF-IDF
vectorizer = TfidfVectorizer()
tfidf_vectors = vectorizer.fit_transform(col2_clean)

# Création d'un modèle de clustering K-means avec 2 clusters
kmeans = KMeans(n_clusters=2, random_state=1000)

# Entraînement du modèle sur les vecteurs TF-IDF
kmeans.fit(tfidf_vectors)

# Obtention des étiquettes de cluster pour chaque commentaire
labels = kmeans.labels_

# Création d'un nouveau dataframe contenant les commentaires et leur étiquette de cluster
clustered_data = pd.DataFrame({'Commentaire': col2_clean, 'Cluster': labels})

# Affichage des commentaires et leur étiquette de cluster
for _, row in clustered_data.iterrows():
    print('Commentaire:', row['Commentaire'])
    print('Cluster:', row['Cluster'])
    print()


Commentaire: peopl would take step back make case wasnt anyon except two peopl situat lump mess take matter hand make kind protest selfish without ration thought investig guy video heavili emot hype want heard get heard press never reason discuss kudo smerconish keep level whole time let masri make fool dare tore citi protest make dishonor entir incid hate way sinc polic brutal becom epidem wish everyon would stop pretend like knew exactli go there measur amount peopl honestli wit incid none us clue way whole issu swung grand juri inform trust major rule right cours action let also thank polic offic america actual serv protect even your bit jerk pull respect job know someon mani peopl go pout held account action peopl hate polic need offic two around emerg
Cluster: 0

Commentaire: law enforc train shoot apprehend train shoot kill thank wilson kill punk bitch
Cluster: 0

Commentaire: dont reckon black live matter banner held white cunt kinda patron iron racist could come somethin better

In [58]:
import pandas as pd

true_labels = dataset['IsToxic']

# Créer une table croisée entre les véritables étiquettes et les étiquettes de cluster prédites
comparaison  = pd.crosstab(true_labels, labels)

# Afficher la table de comparaison
print(comparaison)


col_0      0   1
IsToxic         
False    479  59
True     389  73


IsToxic False (Non toxique) : Il y a 479 occurrences où les données réelles étaient non toxiques et le modèle les a également correctement prédites comme non toxiques (vrais négatifs, TN). Cependant, il y a 59 occurrences où les données réelles étaient non toxiques, mais le modèle les a incorrectement prédites comme toxiques (faux positifs, FP).

IsToxic True (Toxique) : Il y a 73 occurrences où les données réelles étaient toxiques et le modèle les a également correctement prédites comme toxiques (vrais positifs, TP). Cependant, il y a 389 occurrences où les données réelles étaient toxiques, mais le modèle les a incorrectement prédites comme non toxiques (faux négatifs, FN).

#  clustring avec l'algorithme DBSCAN

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.cluster import DBSCAN
import pandas as pd

# Calcul des vecteurs TF-IDF
vectorizer = TfidfVectorizer()
tfidf_vectors = vectorizer.fit_transform(col2_clean)

# Création d'un modèle de clustering DBSCAN
dbscan = DBSCAN(eps=1, min_samples=5)

# Entraînement du modèle sur les vecteurs TF-IDF
dbscan.fit(tfidf_vectors)

# Obtention des étiquettes de cluster pour chaque commentaire
labels = dbscan.labels_

# Création d'un nouveau dataframe contenant les commentaires et leur étiquette de cluster
clustered_data = pd.DataFrame({'Commentaire': col2_clean, 'Cluster': labels})

# Affichage des commentaires et leur étiquette de cluster
for _, row in clustered_data.iterrows():
    print('Commentaire:', row['Commentaire'])
    print('Cluster:', row['Cluster'])
    print()


In [38]:
# Assumant 'dataset' est votre dataframe contenant les véritables étiquettes (IsToxic)
true_labels = dataset['IsToxic']

# Créer une table croisée entre les véritables étiquettes et les étiquettes de cluster prédites
cross_table = pd.crosstab(true_labels, labels)

# Afficher la table croisée pour visualiser la comparaison
print(cross_table)


col_0     -1   0
IsToxic         
False    538   0
True     446  16


"""Réel False : Il y a 538 occurrences où les données réelles étaient classées comme False et le modèle les a également correctement prédites comme False (vrais négatifs, TN).

Réel True : Il y a 7 occurrences où les données réelles étaient classées comme True et le modèle les a également correctement prédites comme True (vrais positifs, TP). Cependant, il y a aussi 455 occurrences où les données réelles étaient classées comme True, mais le modèle les a incorrectement prédites comme False (faux négatifs, FN)."""